examples/custom_render_pass/custom_pass.rs

```diff
@@ -0,0 +1,354 @@
+///         ShaderKind::Fragment,
+///         SourceLanguage::GLSL,
+///         "main",
+///     ).precompile().unwrap();
+/// }
+/// '''
+
+/// Draw triangles.
+#[derive(Clone, Debug, PartialEq, Derivative)]
+#[derivative(Default(bound = ""))]
+pub struct DrawCustomDesc;
+
+impl DrawCustomDesc {
+    /// Create instance of `DrawCustomDesc` render group
+    pub fn new() -> Self {
+        Default::default()
+    }
+}
+
+impl<B: Backend> RenderGroupDesc<B, World> for DrawCustomDesc {
+    fn build(
+        self,
+        _ctx: &GraphContext<B>,
+        factory: &mut Factory<B>,
+        _queue: QueueId,
+        _world: &World,
+        framebuffer_width: u32,
+        framebuffer_height: u32,
+        subpass: hal::pass::Subpass<'_, B>,
+        _buffers: Vec<NodeBuffer>,
+        _images: Vec<NodeImage>,
+    ) -> Result<Box<dyn RenderGroup<B, World>>, failure::Error> {
+        let env = DynamicUniform::new(factory, pso::ShaderStageFlags::VERTEX)?;
+        let vertex = DynamicVertexBuffer::new();
+
+        let (pipeline, pipeline_layout) = build_custom_pipeline(
+            factory,
+            subpass,
+            framebuffer_width,
+            framebuffer_height,
+            vec![env.raw_layout()],
+        )?;
+
+        Ok(Box::new(DrawCustom::<B> {
+            pipeline,
+            pipeline_layout,
+            env,
+            vertex,
+            vertex_count: 0,
+            change: Default::default(),
+        }))
+    }
+}
+
+/// Draws triangles to the screen.
+#[derive(Debug)]
+pub struct DrawCustom<B: Backend> {
+    pipeline: B::GraphicsPipeline,
+    pipeline_layout: B::PipelineLayout,
+    env: DynamicUniform<B, CustomUniformArgs>,
+    vertex: DynamicVertexBuffer<B, CustomArgs>,
+    vertex_count: usize,
+    change: ChangeDetection,
+}
+
+impl<B: Backend> RenderGroup<B, World> for DrawCustom<B> {
+    fn prepare(
+        &mut self,
+        factory: &Factory<B>,
+        _queue: QueueId,
+        index: usize,
+        _subpass: hal::pass::Subpass<'_, B>,
+        world: &World,
+    ) -> PrepareResult {
+        let (triangles,) = <(ReadStorage<'_, Triangle>,)>::fetch(world);
+
+        // Get our scale value
+        let scale = world.read_resource::<CustomUniformArgs>();
+
+        // Write to our DynamicUniform
+        self.env.write(factory, index, scale.std140());
+
+        self.vertex_count = triangles.join().count() * 3;
+
+        // Create an iterator over the Triangle vertices
+        let vertex_data_iter = triangles.join().flat_map(|triangle| triangle.get_args());
+
+        // Write the vector to a Vertex buffer
+        self.vertex.write(
+            factory,
+            index,
+            self.vertex_count as u64,
+            Some(vertex_data_iter.collect::<Box<[CustomArgs]>>()),
+        );
+
+        // Return with we can reuse the draw buffers using the utility struct ChangeDetection
+        self.change.prepare_result(self.vertex_count, false)
```

Comment:
This usage is incorrect. First argument should also be the frame index, second signifies if the change was detected in this frame.

So, first you need to actually detect this.
```rust
let old_vertex_count = self.vertex_count;
self.vertex_count = triangles.join().count() * 3;
let changed = old_vertex_count != self.vertex_count; 
```

Then, you need to submit that as a change for current frame
```rust
self.change.prepare_result(index, changed)
```

Think about it as a replacement for simple `if changed { redraw } else { reuse }`, except taking "frames in flight" into account.

The whole `ChangeDetection` thing is only necessary because of the details of which command buffer is actually resubmitted, as there is a dedicated one per every frame in flight. That means you can only reuse a buffer if a number of consecutive frames were not changed.

Reply:
Ok. I have implemented these and I think I get what they do. We might want to clarify the documentation on these as I obviously didn't grok it the first time.